Parse/src/main/java/com/parse/ParseObjectSubclassingController.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2015-present, Parse, LLC.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+package com.parse;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.Map;
+
+/* package */ class ParseObjectSubclassingController {
+  private static class ObjectSubclassInfo {
+    public final Class<? extends ParseObject> type;
+    public final String className;
+    public final Constructor<? extends ParseObject> constructor;
```

Comment:
I'm not 100% sure about performance. I did some rudimentary searching, but found no conclusive results.

Mostly just caching the constructor for parity with .NET. There's a minute difference as to what kind of exceptions are thrown by `Constructor.newInstance()` vs `Class.newInstance()` as well...